staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go

```diff
@@ -552,16 +652,45 @@ func (alloc *allocator) allocateOne(r deviceIndices) (bool, error) {
 		return alloc.allocateOne(deviceIndices{claimIndex: r.claimIndex + 1})
 	}
 
+	requestKey := requestIndices{claimIndex: r.claimIndex, requestIndex: r.requestIndex, subRequest: r.subRequest, subRequestIndex: r.subRequestIndex}
+	requestData := alloc.requestData[requestKey]
+	// If the request has subRequests, attempt them in order until one can be satisfied
+	// or all have been attempted.
```

Comment:
This doesn't seem to do backtracking. As with claims, requests and devices, `allocateOne` must try to allocate one sub-request, continue with the overall allocation, then try some other sub-request if no solution was found.

Reply:
Let's add the tests, this was a good catch. It's a bit confusing, I recall thinking the original code didn't do backtracking but it does. This should continue to do some backtracking, but you are right that it might not backtrack within the subrequest sequence. That is, we don't restart checking the subrequest at N+1 if we tried with N, succeeded here, but got forced to backtrack for other reasons. The question is - what is the right behavior?

If we backtrack within the subrequest list, what we are saying is "other reasons for failure to schedule take precedence over the prioritized list". I think that's right, since some solution is better than no solution. I think as this is written right now, we pick the first thing in the list, if then we fail for other reasons, we move to the next device. What you are saying is if we fail for other reasons, we should check the next prioritized list entry, before moving to the next device. Is that right?

Wait, I may be a little confused. We are doing a nested iteration through: claims -> requests -> devices, right? So, the nested iteration should now be claims -> requests -> subrequests -> devices, with backtracking at each level>

Reply:
> The question is - what is the right behavior?

If there is some solution, the scheduler should find it. All of the test cases in https://github.com/kubernetes/kubernetes/pull/128586#discussion_r1880440366 have a solution where all requests and subrequests can be allocated, but I think the current code fails to find that solution in the case that requires backtracking.

> So, the nested iteration should now be claims -> requests -> subrequests -> devices, with backtracking at each level?

Correct.

Reply:
> If there is some solution, the scheduler should find it. All of the test cases in https://github.com/kubernetes/kubernetes/pull/128586#discussion_r1880440366 have a solution where all requests and subrequests can be allocated, but I think the current code fails to find that solution in the case that requires backtracking.

Agreed.

Reply:
The code does support backtracking. The recursive call to handle the next claim or request is handled within the for-loop and will continue trying if the first choice doesn't work out.

I added tests for the scenarios mentioned in https://github.com/kubernetes/kubernetes/pull/128586/files#r1880440366 here: https://github.com/mortent/kubernetes/blob/DRAPrioritizedList/staging/src/k8s.io/dynamic-resource-allocation/structured/allocator_test.go#L1752-L1849

Reply:
I have the nagging feeling that `requestIndices.subRequest`, `deviceIndices.subRequest` and this additional `for` loop are not necessary. Finding the allocated subrequest might become a bit slower without that special `alloc.requestData` entry, but I think the could would be simpler and easier to understand.

If you want, then I can try to simplify tomorrow.

Reply:
Sure, I'm happy to take suggestions for improvements.

That the code uses a for loop for trying the different subrequests rather than recursion was a deliberate choice. It just seemed like it would be more understandable. But that was just my opinion.

I went back and forth a bit on how to structure the information in `alloc.requestData` and this seemed like the best option. But I'm not convinced it is the best possible way to do it, so happy to look at alternatives.

Reply:
Thanks @pohly. I will take a look at the suggestions tomorrow and incorporate in the PR.

Reply:
I've incorporated your changes into the PR. I made two adjustments from your branch:

* Changed how the `minDevicesPerRequest` is computed in https://github.com/mortent/kubernetes/blob/DRAPrioritizedList/staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go#L181 and https://github.com/mortent/kubernetes/blob/DRAPrioritizedList/staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go#L194-L196.
* Propagate the value of `allocateSubRequest` in https://github.com/mortent/kubernetes/blob/DRAPrioritizedList/staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go#L757.

It also removes the error in the case of two many devices allocated so that the allocator will continue to look for a solution: https://github.com/mortent/kubernetes/blob/DRAPrioritizedList/staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go#L732-L736. I also added test for this: https://github.com/mortent/kubernetes/blob/DRAPrioritizedList/staging/src/k8s.io/dynamic-resource-allocation/structured/allocator_test.go#L1881

Thanks for the suggestions. I think these are all improvements on the original version.